Inference architecture

tomskip123 opened the pull request.

tomskip123 committed:
started work on openvino integration

tomskip123 committed:
WIP: Instantiated model transposed image to input tensor for yolo

tomskip123 committed:
WIP - gpt 5 output tensor processing code for review

tomskip123 committed:
first revision of openvino class ready - needs testing

tomskip123 committed:
added test image, added ability to select inference lib (WIP)

output is not parsed, but inference is working, on intel GPU its still
around 32ms (on my 12900ks)

Commit:
Refactoring. No longer guesses input tensor format

Commit:
New test for yolo nms

tomskip123 committed:
added onnx inference decode for detr

tomskip123 committed:
added comments for understanding

tomskip123 committed:
why do we divide by 255?

tomskip123 committed:
another gotchya

tomskip123 committed:
Renaming to RFDETRInfer

tomskip123 committed:
WIP - got RFDETR working!

tomskip123 committed:
WIP openvino working with detr

tomskip123 committed:
Merge remote-tracking branch 'origin/main' into inference-architecture

tomskip123 committed:
Add inference pipeline and backend support for Ultralytics YOLO model

- Implemented ModelPipeline class for managing inference workflows.
- Created UltralyticsBackend for interfacing with YOLO model.
- Added preprocessing and postprocessing utilities for handling image data and results.
- Established module structure for inference, preprocessing, and postprocessing components.

tomskip123 committed:
Add warmup inference capability to DetectObjectService and update UltralyticsBackend for device handling

- Implemented warmup inference in DetectObjectService to pre-compile models for GPU/OpenVINO using a specified warmup image.
- Enhanced UltralyticsBackend to allow device selection during prediction, delegating device management to the model itself.

tomskip123 committed:
Update .gitignore and example-config.ini for image file handling and inference settings

- Added support for ignoring .jpg and .png files in .gitignore.
- Updated example-config.ini to change the host to 0.0.0.0, enable verbose logging, specify the device as intel:npu, and set the warmup image path for inference.

tomskip123 committed:
Add OpenVINO and ONNX backend support, update configuration, and enhance preprocessing/postprocessing utilities

- Introduced OpenVINO and ONNX backends for model inference, including necessary classes and methods.
- Updated example-config.ini to include backend and model type options.
- Added preprocessing and postprocessing utilities for YOLO models, including standard and ultralytics-specific functions.
- Removed the Ultralytics backend as it is now replaced by the new structure.
- Enhanced module imports to reflect the new backend architecture.

tomskip123 committed:
Refactor DetectObjectService to support multiple backends and model types

- Enhanced backend selection to include OpenVINO and ONNX alongside Ultralytics.
- Updated preprocessing and postprocessing functions based on selected backend and model type.
- Improved configuration handling for backend and model type, ensuring proper error handling for unsupported combinations.
- Streamlined model loading process to accommodate the new backend architecture.

tomskip123 committed:
Enhance ModelPipeline with timing metrics for preprocessing, inference, and postprocessing

- Added timing measurements for each stage of the inference pipeline.
- Implemented verbose logging to display timing information when enabled.
- Streamlined the call method to improve clarity and maintainability.

tomskip123 committed:
Refactor image handling in DetectObjectService and preprocessing utilities

- Updated DetectObjectService to accept raw RGBA image bytes instead of PNG format.
- Modified warmup inference to load images in RGBA format using OpenCV.
- Renamed preprocessing function from `prepare_png_bytes` to `prepare_rgba_bytes` to reflect the new input format.
- Enhanced input validation to ensure correct image dimensions and types for inference.

tomskip123 committed:
Update gRPC generated files and enhance README for Protobuf generation

- Bump gRPC generated version from 1.75.1 to 1.76.0 in detect_object_pb2_grpc.py.
- Modify RequestPayload to use image_bytes instead of image_png_bytes in detect_object_pb2.py and detect_object_pb2.pyi.
- Adjust serialized field lengths in detect_object_pb2.py for RequestPayload and DetectObject.
- Add instructions to README for generating Protobuf files for Python and Go after modifying .proto definitions.
- Remove outdated Go output files from the repository.

tomskip123 committed:
Optimize YOLO input preprocessing by streamlining image conversion and memory handling

- Refactored `prepare_yolo_input` to use direct indexing for BGR channel extraction, reducing operations and improving performance.
- Pre-allocated output buffer for NCHW format to enhance memory efficiency.
- Updated `prepare_rgba_bytes` to ensure contiguous memory layout while dropping the alpha channel.

tomskip123 committed:
Update .dockerignore, README, and requirements.txt for improved Docker support and documentation

- Expanded .dockerignore to include additional files and directories for Python, IDEs, Docker, CI/CD, and documentation.
- Added a section in README for running the server with Docker Compose, including instructions for switching between backend variants.
- Updated requirements.txt by removing specific model/tooling dependencies to streamline package management.

tomskip123 committed:
Refactor Docker configuration and backend imports for improved flexibility

- Removed the production target from docker-bake.hcl and updated the all group to reflect this change.
- Reorganized docker-compose.yml to enable the OpenVINO NPU variant while commenting out the previous GPU and production configurations.
- Cleaned up the Dockerfile by removing the production build stage and unnecessary package installations.
- Updated server_utils.py and server.py to conditionally import the Ultralytics backend, ensuring compatibility if the package is not available.
- Enhanced OpenvinoBackend to support NPU-specific configurations during model compilation.

tomskip123 committed:
Enhance OpenvinoBackend with NPU driver and compiler version logging

- Added verbose logging to display NPU driver and compiler versions during model compilation.
- Updated comments to clarify NPU-specific configuration handling.

JordanStoker commented on src/inference/preprocess/yolo_standard.py:
```diff
@@ -0,0 +1,48 @@
+    # Pre-allocate output buffer (NCHW format)
+    image_nchw = np.empty((1, 3, h, w), dtype=np.float32)
+    
+    # Vectorized: extract BGR channels, transpose, convert, normalize in minimal ops
```
Why BGR channels? 

Isn't the input tensor (1,3,640,640) where the 3 channels are RGB? 
According to ultralytics AI website
"The input tensor format for YOLO11 is a 4-dimensional array with shape (1, 3, 640, 640), where:

1 is the batch size,
3 is the number of color channels (RGB),
640 and 640 are the height and width of the image, respectively."

From LP we have:
```
func (d *baseMLAxleDetector) setInputTensor(img *image.RGBA) {
	dx, dy := img.Rect.Dx(), img.Rect.Dy()
	data := d.inputTensor.GetData()
	red := data[0*dx*dy : 1*dx*dy]
	green := data[1*dx*dy : 2*dx*dy]
	blue := data[2*dx*dy : 3*dx*dy]

	// Loop over all pixels
	idx := 0
	for y := img.Rect.Min.Y; y < img.Rect.Max.Y; y++ {
		for x := img.Rect.Min.X; x < img.Rect.Max.X; x++ {
			// Normalize pixel values to [0, 1] range.
			i := img.PixOffset(x, y)    // should be optimised by increment i by 4 in every loop.
			s := img.Pix[i : i+3 : i+3] // Small cap improves performance, see https://golang.org/issue/27857
			red[idx] = float32(s[0]) / 255.0
			green[idx] = float32(s[1]) / 255.0
			blue[idx] = float32(s[2]) / 255.0
			idx++
		}
	}
}
```

JordanStoker commented on src/inference/preprocess/yolo_standard.py:
```diff
@@ -0,0 +1,48 @@
+    h, w = orig_shape
+    
+    # Pre-allocate output buffer (NCHW format)
+    image_nchw = np.empty((1, 3, h, w), dtype=np.float32)
```
Can we cache this? 

First thought is to create a YOLO class with prepare_input and prepare_output functions. The class can cache variables

Since we may have multiple LP's requesting inference, we'd want this to be safe for concurrent usage. Simplest may be having a set of objects per client?

JordanStoker commented on server.py:
```diff
@@ -13,77 +12,176 @@
+            )
+
+            pipeline_kwargs = {
+                "imgsz": (640, 640),
```
Should this be here? the `640x640`
If we use a different model with different input tensor shape then this would be incorrect.

Should it be something the Model type specific class handles (YOLO, RF-DETR etc...)?

tomskip123 committed:
Update Dockerfile to use Ubuntu 24.04 and streamline Python package installations

- Changed base image from Ubuntu 22.04 to 24.04.
- Updated Python installation to use the latest version and removed specific version dependencies.
- Added `--break-system-packages` flag to pip install commands to address compatibility issues.
- Cleaned up package installations for OpenVINO and ultralytics to enhance image efficiency.

tomskip123 committed:
Update version numbers in configuration files to 0.0.1

- Updated app_version in docker-bake.hcl and server.py to reflect the new version.
- Modified image tags in docker-compose.yml for the inference-server to version 0.0.1 for all variants.

tomskip123 committed:
setting default image service to the cpu service, changes should not be commited to this file.

tomskip123 committed:
Enhance backend imports and error handling for OpenVINO support in inference service

JordanStoker commented on server.py:
```diff
@@ -13,77 +12,178 @@
+
+        # Warmup inference to pre-compile model for GPU/OpenVINO
+        warmup_enabled = bool(int(self.config.get("WarmupInference", "0")))
+        warmup_image_path = self.config.get("WarmupImagePath", "tests/img_test.png")
```
Is this needed? Can't we just warm up the model with a random/empty tensor? Was able to do so in Onnx with LP

tomskip123 committed:
added image saving pipeline

tomskip123 committed:
Add configuration option for image saving directory and update save logic

tomskip123 committed:
Refactor image saving directory configuration and remove unused inference files
